e2e/testcases/namespace_strategy_test.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/applier"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/declared"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/testing/fake"
+	"sigs.k8s.io/cli-utils/pkg/common"
+)
+
+func TestNamespaceStrategy(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.OverrideAPI, ntopts.Unstructured)
+
+	rootSyncNN := nomostest.RootSyncNN(configsync.RootSyncName)
+	rootReconcilerNN := core.RootReconcilerObjectKey(rootSyncNN.Name)
+	rootSync := fake.RootSyncObjectV1Alpha1(rootSyncNN.Name)
+	// set the NamespaceStrategy to explicit
+	nt.MustMergePatch(rootSync, `{"spec": {"override": {"namespaceStrategy": "explicit"}}}`)
+	err := nt.Watcher.WatchObject(
+		kinds.Deployment(), rootReconcilerNN.Name, rootReconcilerNN.Namespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(
+				reconcilermanager.Reconciler,
+				reconcilermanager.NamespaceStrategy,
+				string(configsync.NamespaceStrategyExplicit),
+			),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	// add a resource for which the namespace is not declared/created
+	fooNamespace := fake.NamespaceObject("foo-implicit")
+	cm1 := fake.ConfigMapObject(core.Name("cm1"), core.Namespace(fooNamespace.Name))
+	nt.Must(nt.RootRepos[configsync.RootSyncName].Add("acme/cm1.yaml", cm1))
+	nt.Must(nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add cm1"))
+	// check for error
+	nt.WaitForRootSyncSyncError(rootSyncNN.Name, applier.ApplierErrorCode,
```

Comment:
Should we perhaps also test the use case where this error recovers automatically when the namespace is created by another client?

Another possible permutation would be to have one RootSync block/error until a second RootSync provisions the namespace it needs.

I don't know if we need all these cases coved in e2e, but they seem pretty common.

Reply:
Seems reasonable - I'm thinking I can add another testcase which covers the following behavior:

- sync-1 creates a resource with namespaceStrategy=explicit, and expects an error
- sync-2 creates the namespace
- assert that both sync-1 and sync-2 successfully sync, and that the namespace is managed by sync-1

Reply:
added another testcase which approximates the scenario described in the original issue